pkg/cmd/server/config.go

```diff
@@ -0,0 +1,250 @@
+		KubernetesAddr:       flagtypes.Addr{DefaultScheme: "https", DefaultPort: 8443}.Default(),
+		PortalNet:            flagtypes.DefaultIPNet("172.30.17.0/24"),
+		MasterPublicAddr:     flagtypes.Addr{Value: "localhost:8443", DefaultScheme: "https", DefaultPort: 8443, AllowPrefix: true}.Default(),
+		KubernetesPublicAddr: flagtypes.Addr{Value: "localhost:8443", DefaultScheme: "https", DefaultPort: 8443, AllowPrefix: true}.Default(),
+
+		ImageTemplate: variable.NewDefaultImageTemplate(),
+
+		Hostname: hostname,
+		NodeList: flagtypes.StringList{"127.0.0.1"},
+	}
+}
+
+// GetMasterAddress checks for an unset master address and then attempts to use the first
+// public IPv4 non-loopback address registered on this host. It will also update the
+// EtcdAddr after if it was not provided.
+// TODO: make me IPv6 safe
+func (cfg Config) GetMasterAddress() (*url.URL, error) {
+	if cfg.MasterAddr.Provided {
+		return cfg.MasterAddr.URL, nil
+	}
+
+	if cfg.StartMaster {
+		// If the user specifies a bind address, and the master is not provided, use the bind port by default
+		port := cfg.MasterAddr.Port
+		if cfg.BindAddr.Provided {
+			port = cfg.BindAddr.Port
+		}
+
+		// If the user specifies a bind address, and the master is not provided, use the bind scheme by default
+		scheme := cfg.MasterAddr.URL.Scheme
+		if cfg.BindAddr.Provided {
+			scheme = cfg.BindAddr.URL.Scheme
+		}
+
+		// use the default ip address for the system
+		addr, err := util.DefaultLocalIP4()
+		if err != nil {
+			return nil, fmt.Errorf("Unable to find the public address of this master: %v", err)
+		}
+
+		masterAddr := scheme + "://" + net.JoinHostPort(addr.String(), strconv.Itoa(port))
+		return url.Parse(masterAddr)
+	}
+
+	// if we didn't specify and we aren't starting the master, read .kubeconfig to locate the master
+	// TODO client config currently doesn't let you override the defaults
+	// so it is defaulting to https://localhost:8443 for MasterAddr if
+	// it isn't set by --master or --kubeconfig
+	config, err := cfg.ClientConfig.ClientConfig()
+	if err != nil {
+		return nil, err
+	}
+	return url.Parse(config.Host)
+}
+
+func (cfg Config) GetMasterPublicAddress() (*url.URL, error) {
+	if cfg.MasterPublicAddr.Provided {
+		return cfg.MasterPublicAddr.URL, nil
+	}
+
+	return cfg.GetMasterAddress()
+}
+
+func (cfg Config) GetEtcdAddress() (*url.URL, error) {
+	if cfg.EtcdAddr.Provided {
+		return cfg.EtcdAddr.URL, nil
+	}
+
+	// Etcd should be reachable on the same address that the master is (for simplicity)
+	masterAddr, err := cfg.GetMasterAddress()
+	if err != nil {
+		return nil, err
+	}
+
+	etcdAddr := net.JoinHostPort(getHost(*masterAddr), strconv.Itoa(cfg.EtcdAddr.DefaultPort))
+	return url.Parse("http://" + etcdAddr)
+}
+
+func (cfg Config) GetKubernetesAddress() (*url.URL, error) {
+	if cfg.KubernetesAddr.Provided {
+		return cfg.KubernetesAddr.URL, nil
+	}
+
+	return cfg.GetMasterAddress()
+}
+
+func (cfg Config) GetKubernetesPublicAddress() (*url.URL, error) {
```

Comment:
should follow this chain: KubernetesPublicAddress, KubernetesAddress, MasterPublicAddress, MasterAddress